# BT4222 Assignment 1
By Teo Zhi Feng

# 1 Data Exploration

### 1.1 Load datasets

In [1]:
import pandas as pd
from numpy import mean
from collections import Counter

In [2]:
df_train = pd.read_csv('/Users/BigTeo/Desktop/BT4222/assignment1/train.csv')
df_train.head(3)

index  age    race    sex      education   f1    f2   f3    f4   f5  \
0      1   26   White   Male   Some-college  a_2   b_4  c_0  d_39  e_4   
1      2   35   White   Male   Some-college  a_2   b_3  c_0  d_39  e_4   
2      3   61   White   Male        HS-grad  a_2  b_12  c_0  d_39  e_6   

       f6  f7  f8    f9  f10  y  
0   53833  10   0     0   42  0  
1   67728  10   0  2051   45  0  
2  268831   9   0     0   53  0

In [3]:
df_test = pd.read_csv('/Users/BigTeo/Desktop/BT4222/assignment1/test.csv')
df_test.head(3)

index  age    race    sex      education   f1   f2   f3    f4   f5      f6  \
0      1   39   White   Male        HS-grad  a_2  b_3  c_0  d_39  e_4  366757   
1      2   56   White   Male   Some-college  a_2  b_0  c_0  d_27  e_0  275943   
2      3   17   White   Male           12th  a_4  b_0  c_3  d_39  e_0  103810   

   f7  f8  f9  f10  
0   9   0   0   35  
1  10   0   0   40  
2   8   0   0   40

### 1.2 Check distribution of independent variable (y)
##### Insight: Train dataset is imbalanced where the ratio of y=0 to y=1 is approximately 3:1.

In [4]:
Counter(df_train['y'])

Counter({0: 20194, 1: 6367})

In [5]:
ratio = len(df_train[ df_train['y'] == 0]) / len(df_train[ df_train['y'] == 1])
print('Ratio of df_train[\'y\']=0 to df_train[\'y\']=1: %.2f' % (ratio))

Ratio of df_train['y']=0 to df_train['y']=1: 3.17


### 1.3 Check type of dependent variables
##### Insight: As 'race', 'sex', 'education', 'f1', 'f2', 'f3', 'f4', 'f5' are categorical data, there is a need to convert  into dummy variables.

In [6]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26561 entries, 0 to 26560
Data columns (total 16 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   index      26561 non-null  int64 
 1   age        26561 non-null  int64 
 2   race       26561 non-null  object
 3   sex        26561 non-null  object
 4   education  26561 non-null  object
 5   f1         26561 non-null  object
 6   f2         26561 non-null  object
 7   f3         26561 non-null  object
 8   f4         26561 non-null  object
 9   f5         26561 non-null  object
 10  f6         26561 non-null  int64 
 11  f7         26561 non-null  int64 
 12  f8         26561 non-null  int64 
 13  f9         26561 non-null  int64 
 14  f10        26561 non-null  int64 
 15  y          26561 non-null  int64 
dtypes: int64(8), object(8)
memory usage: 3.2+ MB


In [7]:
df_train.describe()

index           age            f6            f7            f8  \
count  26561.000000  26561.000000  2.656100e+04  26561.000000  26561.000000   
mean   13281.000000     38.610933  1.902328e+05     10.073981   1070.548662   
std     7667.644586     13.671748  1.056464e+05      2.573905   7320.365123   
min        1.000000     17.000000  1.228500e+04      1.000000      0.000000   
25%     6641.000000     28.000000  1.183990e+05      9.000000      0.000000   
50%    13281.000000     37.000000  1.783830e+05     10.000000      0.000000   
75%    19921.000000     48.000000  2.376700e+05     12.000000      0.000000   
max    26561.000000     90.000000  1.484705e+06     16.000000  99999.000000   

                 f9           f10             y  
count  26561.000000  26561.000000  26561.000000  
mean      87.575129     40.488573      0.239712  
std      404.081775     12.358773      0.426916  
min        0.000000      1.000000      0.000000  
25%        0.000000     40.000000      0.000000  
50%        0.000000     40.000000      0.000000  
75%        0.000000     45.000000      0.000000  
max     4356.000000     99.000000      1.000000

# 2 Data Preparation

### 2.1 Encode categorical variables in df_train

In [8]:
df_train = pd.get_dummies(df_train, columns = ['race', 'sex', 'education', 'f1', 'f2', 'f3', 'f4', 'f5'])
df_train.head(3)

index  age      f6  f7  f8    f9  f10  y  race_ Amer-Indian-Eskimo  \
0      1   26   53833  10   0     0   42  0                         0   
1      2   35   67728  10   0  2051   45  0                         0   
2      3   61  268831   9   0     0   53  0                         0   

   race_ Asian-Pac-Islander  ...  f4_d_9  f5_e_0  f5_e_1  f5_e_2  f5_e_3  \
0                         0  ...       0       0       0       0       0   
1                         0  ...       0       0       0       0       0   
2                         0  ...       0       0       0       0       0   

   f5_e_4  f5_e_5  f5_e_6  f5_e_7  f5_e_8  
0       1       0       0       0       0  
1       1       0       0       0       0  
2       0       0       1       0       0  

[3 rows x 109 columns]

### 2.2 Create X and y variables for df_train

In [9]:
X_train = df_train.drop('y', axis = 1)
y_train = df_train['y']

### 2.3 Encode categorical variables in df_test

In [10]:
X_test = pd.get_dummies(df_test, columns = ['race', 'sex', 'education', 'f1', 'f2', 'f3', 'f4', 'f5'])
X_test.head(3)

index  age      f6  f7  f8  f9  f10  race_ Amer-Indian-Eskimo  \
0      1   39  366757   9   0   0   35                         0   
1      2   56  275943  10   0   0   40                         0   
2      3   17  103810   8   0   0   40                         0   

   race_ Asian-Pac-Islander  race_ Black  ...  f4_d_9  f5_e_0  f5_e_1  f5_e_2  \
0                         0            0  ...       0       0       0       0   
1                         0            0  ...       0       1       0       0   
2                         0            0  ...       0       1       0       0   

   f5_e_3  f5_e_4  f5_e_5  f5_e_6  f5_e_7  f5_e_8  
0       0       1       0       0       0       0  
1       0       0       0       0       0       0  
2       0       0       0       0       0       0  

[3 rows x 109 columns]

### 2.4 Drop additional columns in X_test
Identify additional columns and remove.

In [11]:
train_cols = X_train.columns
test_cols = X_test.columns

not_in_test_cols = test_cols.difference(train_cols)
not_in_test_cols

Index(['f4_d_15'], dtype='object')

In [12]:
X_test = X_test.drop(columns=[not_in_test_cols[0]], axis=1)

In [13]:
print(X_train.shape)
print(X_test.shape)

(26561, 108)
(6000, 108)


# 3 Model Evaluation
Since dataset is imbalanced, stratified k-fold cross-validation is used to preserve the class distribution in train and validation dataset. To prevent data leakage, a pipeline will be set up to perform data preparation and confuct grid search for hyperparameter tuning prior to running the model. 

In [14]:
from sklearn.model_selection import cross_validate, StratifiedKFold, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

In [15]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.gaussian_process.kernels import RBF
from sklearn.ensemble import AdaBoostClassifier

### 3.1 Setting up

In [16]:
import warnings
warnings.filterwarnings('ignore')

In [17]:
metrics = ['accuracy', 'f1_weighted', 'roc_auc']

In [18]:
# define dataframe to store results
results = pd.DataFrame(columns=['Model', 'Accuracy', 'F1', 'ROC AUC'])

In [19]:
# define evaluation procedure
kfold = StratifiedKFold(n_splits=3, random_state=1, shuffle=True)

In [20]:
def run_model(model, parameters):
    # set up pipeline
    pipeline = Pipeline(steps=[
        ('scaler', StandardScaler()),
        ('model', model)
    ])
    
    # grid search
    search = GridSearchCV(estimator=pipeline,
                            param_grid=parameters,
                            cv=kfold,
                            scoring=metrics,
                            refit='f1_weighted',
                            return_train_score=True,
                            verbose=3)
    search.fit(X_train, y_train)

    # store results in dataframe
    global results
    results = results.append({
        'Model': model,
        'Accuracy': search.cv_results_['mean_test_accuracy'].max(),
        'F1': search.cv_results_['mean_test_f1_weighted'].max(),
        'ROC AUC': search.cv_results_['mean_test_roc_auc'].max()}, 
        ignore_index=True)

    return search

### 3.2 Logistic Regression

In [21]:
params_logreg = {
    'model__C': [0.001,0.01,0.1,1,10,100,1000]
}

In [22]:
model_logreg = run_model(LogisticRegression(), params_logreg)

Fitting 3 folds for each of 7 candidates, totalling 21 fits
[CV 1/3] END model__C=0.001; accuracy: (train=0.844, test=0.841) f1_weighted: (train=0.834, test=0.829) roc_auc: (train=0.901, test=0.899) total time=   0.1s
[CV 2/3] END model__C=0.001; accuracy: (train=0.843, test=0.844) f1_weighted: (train=0.833, test=0.833) roc_auc: (train=0.900, test=0.899) total time=   0.1s
[CV 3/3] END model__C=0.001; accuracy: (train=0.843, test=0.841) f1_weighted: (train=0.832, test=0.831) roc_auc: (train=0.901, test=0.896) total time=   0.1s
[CV 1/3] END model__C=0.01; accuracy: (train=0.852, test=0.850) f1_weighted: (train=0.847, test=0.842) roc_auc: (train=0.908, test=0.906) total time=   0.1s
[CV 2/3] END model__C=0.01; accuracy: (train=0.852, test=0.853) f1_weighted: (train=0.846, test=0.847) roc_auc: (train=0.908, test=0.906) total time=   0.1s
[CV 3/3] END model__C=0.01; accuracy: (train=0.853, test=0.846) f1_weighted: (train=0.847, test=0.841) roc_auc: (train=0.909, test=0.904) total time=   

In [23]:
print('LOGISTIC REGRESSION')
print('Best parameter: ', model_logreg.best_params_)
print('Best F1 score: ', model_logreg.best_score_)
print('-----------------------------------------------------------------------------')
print(results)

LOGISTIC REGRESSION
Best parameter:  {'model__C': 1}
Best F1 score:  0.8459422262339197
-----------------------------------------------------------------------------
                  Model  Accuracy        F1   ROC AUC
0  LogisticRegression()  0.851286  0.845942  0.906375


### 3.3 Adaboost

In [24]:
params_adaboost = {
    'model__n_estimators': [10, 50, 100, 500],
    'model__learning_rate': [0.0001, 0.001, 0.01, 0.1, 1.0]
}

In [25]:
model_adaboost = run_model(AdaBoostClassifier(), params_adaboost)

Fitting 3 folds for each of 20 candidates, totalling 60 fits
[CV 1/3] END model__learning_rate=0.0001, model__n_estimators=10; accuracy: (train=0.760, test=0.760) f1_weighted: (train=0.657, test=0.657) roc_auc: (train=0.760, test=0.759) total time=   0.3s
[CV 2/3] END model__learning_rate=0.0001, model__n_estimators=10; accuracy: (train=0.760, test=0.760) f1_weighted: (train=0.657, test=0.657) roc_auc: (train=0.758, test=0.763) total time=   0.3s
[CV 3/3] END model__learning_rate=0.0001, model__n_estimators=10; accuracy: (train=0.760, test=0.760) f1_weighted: (train=0.657, test=0.657) roc_auc: (train=0.761, test=0.757) total time=   0.3s
[CV 1/3] END model__learning_rate=0.0001, model__n_estimators=50; accuracy: (train=0.760, test=0.760) f1_weighted: (train=0.657, test=0.657) roc_auc: (train=0.760, test=0.759) total time=   1.1s
[CV 2/3] END model__learning_rate=0.0001, model__n_estimators=50; accuracy: (train=0.760, test=0.760) f1_weighted: (train=0.657, test=0.657) roc_auc: (train=0.

[CV 1/3] END model__learning_rate=0.1, model__n_estimators=100; accuracy: (train=0.854, test=0.855) f1_weighted: (train=0.843, test=0.843) roc_auc: (train=0.909, test=0.908) total time=   2.1s
[CV 2/3] END model__learning_rate=0.1, model__n_estimators=100; accuracy: (train=0.855, test=0.857) f1_weighted: (train=0.844, test=0.846) roc_auc: (train=0.908, test=0.909) total time=   2.1s
[CV 3/3] END model__learning_rate=0.1, model__n_estimators=100; accuracy: (train=0.857, test=0.853) f1_weighted: (train=0.845, test=0.843) roc_auc: (train=0.910, test=0.906) total time=   2.1s
[CV 1/3] END model__learning_rate=0.1, model__n_estimators=500; accuracy: (train=0.862, test=0.861) f1_weighted: (train=0.856, test=0.852) roc_auc: (train=0.920, test=0.917) total time=  10.3s
[CV 2/3] END model__learning_rate=0.1, model__n_estimators=500; accuracy: (train=0.861, test=0.862) f1_weighted: (train=0.854, test=0.855) roc_auc: (train=0.919, test=0.918) total time=   9.9s
[CV 3/3] END model__learning_rate=0

In [26]:
print('ADABOOST')
print('Best parameter: ', model_adaboost.best_params_)
print('Best F1 score: ', model_adaboost.best_score_)
print('-----------------------------------------------------------------------------')
print(results)

ADABOOST
Best parameter:  {'model__learning_rate': 1.0, 'model__n_estimators': 500}
Best F1 score:  0.8651486632321271
-----------------------------------------------------------------------------
                  Model  Accuracy        F1   ROC AUC
0  LogisticRegression()  0.851286  0.845942  0.906375
1  AdaBoostClassifier()  0.869433  0.865149  0.924444


### 3.4 Support Vector Machine (SVM)
Given the large number of variables, gridsearch algorithm will not be run on SVM.

In [27]:
from sklearn.model_selection import cross_val_score

In [28]:
X_train = StandardScaler().fit_transform(X_train)

In [29]:
linear_svc = SVC(kernel="linear", C=1, gamma=0.1)
rbf_svc = SVC(kernel='rbf', C=1, gamma=0.1)
poly_svc = SVC(kernel='poly', C=1, gamma=0.1)
sigmoid_svc = SVC(kernel='sigmoid', C=1, gamma=0.1)

all_svm = [linear_svc, rbf_svc, poly_svc, sigmoid_svc]

In [30]:
for model in all_svm: 
    scores = cross_validate(model, X_train, y_train, scoring=metrics, cv=kfold, n_jobs=-1)
    
    results = results.append({
    'Model': model,
    'Accuracy': mean(scores['test_accuracy']),
    'F1': mean(scores['test_f1_weighted']),
    'ROC AUC': mean(scores['test_roc_auc'])},
    ignore_index=True)
    
    print(str(model))
    print('F1 score: ', mean(scores['test_roc_auc']))
    print('-----------------------------------------------------------------------------')
    print(results)
    print()

SVC(C=1, gamma=0.1, kernel='linear')
F1 score:  0.9038970069618338
-----------------------------------------------------------------------------
                                  Model  Accuracy        F1   ROC AUC
0                  LogisticRegression()  0.851286  0.845942  0.906375
1                  AdaBoostClassifier()  0.869433  0.865149  0.924444
2  SVC(C=1, gamma=0.1, kernel='linear')  0.850759  0.844116  0.903897

SVC(C=1, gamma=0.1)
F1 score:  0.8199077396945736
-----------------------------------------------------------------------------
                                  Model  Accuracy        F1   ROC AUC
0                  LogisticRegression()  0.851286  0.845942  0.906375
1                  AdaBoostClassifier()  0.869433  0.865149  0.924444
2  SVC(C=1, gamma=0.1, kernel='linear')  0.850759  0.844116  0.903897
3                   SVC(C=1, gamma=0.1)  0.824291  0.809875  0.819908

SVC(C=1, gamma=0.1, kernel='poly')
F1 score:  0.8331683895424024
------------------------------

# 5 Summary of Performance
Since the dataset is imbalanced, comparing the performances of models using accuracy may be misleading. Hence, F1-score and Area Under The Curve Receiver Operating Characteristics (AUC ROC).

In [31]:
results

Model  Accuracy        F1   ROC AUC
0                   LogisticRegression()  0.851286  0.845942  0.906375
1                   AdaBoostClassifier()  0.869433  0.865149  0.924444
2   SVC(C=1, gamma=0.1, kernel='linear')  0.850759  0.844116  0.903897
3                    SVC(C=1, gamma=0.1)  0.824291  0.809875  0.819908
4     SVC(C=1, gamma=0.1, kernel='poly')  0.811490  0.810499  0.833168
5  SVC(C=1, gamma=0.1, kernel='sigmoid')  0.772637  0.772437  0.753377

As can be seen from the summary above, the Adaboost algorithm performed the best in terms of F1-score (86.5%) and ROC AUC (92.4%). Hence, Adaboost will be used to make predictions on test dataset.

# 6 Run Adaboost on test dataset

### 6.1 Use best adaboost model to predict y

In [32]:
final_model = model_adaboost.best_estimator_

In [33]:
predictions = final_model.predict(X_test)

In [34]:
df_test['y'] = predictions

### 6.2 Check output

In [35]:
df_test.head(3)

index  age    race    sex      education   f1   f2   f3    f4   f5      f6  \
0      1   39   White   Male        HS-grad  a_2  b_3  c_0  d_39  e_4  366757   
1      2   56   White   Male   Some-college  a_2  b_0  c_0  d_27  e_0  275943   
2      3   17   White   Male           12th  a_4  b_0  c_3  d_39  e_0  103810   

   f7  f8  f9  f10  y  
0   9   0   0   35  0  
1  10   0   0   40  0  
2   8   0   0   40  0

In [36]:
Counter(df_test['y'])

Counter({0: 4835, 1: 1165})

In [37]:
ratio = len(df_test[df_test['y'] == 0]) / len(df_test[df_test['y'] == 1])
print('Ratio of df_test[\'y\']=0 to df_test[\'y\']=1: %.2f' % (ratio))

Ratio of df_test['y']=0 to df_test['y']=1: 4.15


### 6.3 Export results to .csv

In [38]:
df_test.to_csv("A1_A0203989N.csv", index=False)